build.assets/Dockerfile-multiarch-base

```diff
@@ -0,0 +1,127 @@
+# It guarantees that we would use the same GCC version for all builds, 
+# and we would be able to upgrade GCC to a newer version if required.
+# #############################################################################
+
+FROM centos:7 AS base
+
+# Automatically supplied by the Docker buildkit
+ARG TARGETARCH
+
+ARG GCC_VERSION=10.4.0
+ARG CMAKE3_VERSION=3.12.3
+
+# #############################################################################
+# Platform-specific customisation.
+# #############################################################################
+
+## ARM 32 #####################################################################
+FROM base AS platform-setup-arm
+ENV GCC_BUILD_FLAGS="--build arm-unknown-linux-gnueabi --with-float=hard --with-mode=arm"
+ENV GOLANG_ARCH=armv6l
+ENV RUST_ARCH=armv7-unknown-linux-gnueabihf
+
+ARG YUM_UPDATE_FLAGS=""
+RUN echo "armhfp" > /etc/yum/vars/basearch && \
+    echo "armv7hl" > /etc/yum/vars/arch && \
+    echo "armv7hl-redhat-linux-gnu" > /etc/rpm/platform
+
+## ARM 64 #####################################################################
+FROM base AS platform-setup-arm64
+ENV GCC_BUILD_FLAGS="--build aarch64-unknown-linux-gnu"
+ENV GOLANG_ARCH=arm64
+ENV RUST_ARCH=aarch64-unknown-linux-gnu
+
+# Installing the kerenel packages causes the update to hang on aarch64, so we
+# skip upgrading them.
+ARG YUM_UPDATE_FLAGS="--exclude kernel-*"
+RUN echo "aarch64-redhat-linux-gnu" > /etc/rpm/platform
+
+## 386 ########################################################################
+FROM base AS platform-setup-386
+ENV GCC_BUILD_FLAGS="--build i386-unknown-linux-gnu"
+ENV GOLANG_ARCH=386
+ENV RUST_ARCH=i686-unknown-linux-gnu
+
+## AMD 64 #####################################################################
+FROM base AS platform-setup-amd64
+ENV GCC_BUILD_FLAGS="--build x86_64-unknown-linux-gnu"
+ENV GOLANG_ARCH=amd64
+ENV RUST_ARCH=x86_64-unknown-linux-gnu
+
+## GCC built from sources #####################################################
+FROM platform-setup-$TARGETARCH AS gcc
+
+ENV LANGUAGE=en_US.UTF-8 \
+    LANG=en_US.UTF-8 \
+    LC_ALL=en_US.UTF-8 \
+    LC_CTYPE=en_US.UTF-8
+
+RUN yum update -y $YUM_UPDATE_FLAGS && \
+    yum install -y     \
+        git \
+        gcc \
+        gcc-c++ \
+        make \
+        pam-devel \
+        perl-IPC-Cmd \
+        tree \
+        which \
+        zip \
+        zlib-static \
+        automake \
+        libtool \
+        flex \        
+        autoconf-archive \
+        libudev-devel \
+        make \
+        wget \
+        gcc \
+        gcc-c++ \
+        glibc-static \
+        glibc-devel \
+        libstdc++-static \
+        libstdc++-devel \
+        bzip2 \
+        file \
+        binutils \
+        zlib-devel \
+        elfutils-libelf-devel-static \
+        texinfo && \
+    yum clean all && \
+    localedef -c -i en_US -f UTF-8 en_US.UTF-8
+
+# Compile & install GCC
```

Comment:
From what I see, we're compiling GCC 10 against the dependencies used to build GCC 4.8. Is this an officially supported combination, or does it just happen to work? I'm talking about MPFR, MPC, etc.

Reply:
Yes. It depends. Some deps require newer GCC to build - but as far as I understood, we meet minimal versions for our case (4.8.3+).